include/alp/smf.hpp

```diff
@@ -27,27 +27,179 @@
+			 * lowercase coefficients are run-time constant,
+			 * and x and y are variables.
+			 * All coefficients, variables are integers and all operations are integer
+			 * operations.
+			 *
+			 * The purpose of compile-time constant coefficients is to allow compile-time
+			 * optimizations for zero factors.
+			 */
+			template< size_t Ax2, size_t Ay2, size_t Axy, size_t Ax, size_t Ay, size_t A0, size_t Denominator >
+			struct BivariateQuadratic {
 
-                /** Maps logical to physical coordinate
-                 * */
-                virtual size_t map(size_t i) = 0;
+				static_assert( Denominator != 0, "Denominator cannot be zero (division by zero).");
 
-                /** Returns the physical dimension of the container needed to
-                 * store all elements
-                 * */
-                virtual size_t allocSize() const = 0;
+				const size_t ax2, ay2, axy, ax, ay, a0;
 
-        };
+				BivariateQuadratic(
+					const size_t ax2, const size_t ay2, const size_t axy,
+					const size_t ax, const size_t ay,
+					const size_t a0 ) :
+					ax2( ax2 ), ay2( ay2 ), axy( axy ),
+					ax( ax ), ay( ay ),
+					a0( a0 ) {}
+
+				size_t f( const size_t x, const size_t y ) const {
+					return (Ax2 * ax2 * x * x +
+						Ay2 * ay2 * y * y +
+						Axy * axy * x * y +
+						Ax * ax * x +
+						Ay * ay * y +
+						A0 * a0) / Denominator;
+				}
+
+			}; // BivariateQuadratic
+
+		};
+
+		typedef polynomials::BivariateQuadratic< 0, 0, 0, 0, 0, 0, 1 > None_t;
+		typedef polynomials::BivariateQuadratic< 0, 0, 0, 1, 1, 0, 1 > Full_t;
+		typedef polynomials::BivariateQuadratic< 0, 0, 0, 0, 0, 0, 1 > Packed_t; // TODO
+		typedef polynomials::BivariateQuadratic< 0, 0, 0, 0, 0, 0, 1 > Band_t; // TODO
+
+		/**
+		 * Creates a polynomial object representing desired storage mapping function.
+		 */
+		struct SMFFactory {
+
+			/**
+			 * For row-major storage, N represents number of columns.
+			 * For column-major storage, N represents number of rows.
+			 */
+			static Full_t Full( size_t N ) {
+				return Full_t( 0, 0, 0, N, 1, 0 );
+			}
+
+			/**
+			 * For triangular and symmetric structures (assuming square matrices)
+			 */
+			static Packed_t Packed( size_t N ) {
+				(void)N;
+				return Packed_t( /* TODO */ 0, 0, 0, 0, 0, 0 );
+			}
+
+			/**
+			 * For Banded structures
+			 */
+			static Band_t Band( size_t rows, size_t cols, size_t kl, size_t ku ) {
+				(void)rows;
+				(void)cols;
+				(void)kl;
+				(void)ku;
+				return Band_t( /* TODO */ 0, 0, 0, 0, 0, 0 );
+			}
+
+		}; // SMFFactory
+
+		/**
+		 * Provides a type of composed Access Mapping Function
+		 * expressed as a BivariateQuadratic polynomial depending
+		 * on the types of the IMFs and the SMF.
+		 */
+		template< typename ImfR, typename ImfC, typename Smf >
+		struct Composition {
+			typedef polynomials::BivariateQuadratic< 1, 1, 1, 1, 1, 1, 1 > type;
+		};
+
+		template<>
+		struct Composition< imf::Strided, imf::Strided, Full_t > {
```

Comment:
I think this could also be generalized. Could we combine the template coefficients of the source polynomials to determine the final composition?